### ISSUE TITLE
Bug: submit -> submitSuccess -> reinitialize -> form.pristine == true => *edit_form* -> form.pristine == TRUE

### ISSUE THREAD

**Issue by rasdaniil** [erikras/redux-form#3729]
A Possible Bug with global form pristine field

### What is the current behavior?
Given reduxForm with
1) enableReinitialize = true
2) keepDirtyOnReinitialize = false (default)
2) onSubmitSuccess(res, dispatch, props) {
  dispatch(initialize(props.form.name, props.form.values)
}

After the form was successfully submitted, it reinitializes and pristine becomes true again (very good).
However, after I make any alteration to the form pristine keeps staying true.

Here is what goes on in the debugger:

```
// createReduxForm.js: 929
const shouldUpdateInitialValues = enableReinitialize (true) && initialized(true) && deepEqual(initialValues, stateInitial) (deepEqual returns false)
// true
```
A combination of `shouldUpdateInitialValues == true` and `keepDirtyOnReinitialize == false` always produces `pristine = true` a few lines below. 

If I set `keepDirtyOnReinitialize` to true, the pristine prop gets evaluated correctly.

I wasn't able to grasp from the documentation how `keepDirtyOnReinitialize` affects global pristine, because it only refers to fields' pristine.

Thanks for the awesome piece of work.

### What is the expected behavior?

I expect pristine to become false after form data is changed.

**Comment by dmitrykrylov** [comment]
I have the same issue. It's a piece of my code: 
```
Form = reduxForm({
    onSubmitSuccess: (result, dispatch, props) => {
        dispatch(initialize(
            props.form,
            props.values,
            { keepSubmitSucceeded: true }
        ));
    }
})(Form);
```

And after `SET_SUBMIT_SUCCEEDED => INITIALIZE  => CHANGE` I have `pristine === true`

**Comment by iamdanthedev** [comment]
@dmitrykrylov 
Hello! I don't  remember really, I suppose I would have posted it here. Could you try setting `keepDirtyOnReinitialize` to true? Maybe that was the thing.. Thanks

**Comment by dmitrykrylov** [comment]
@iamdanthedev I actually want to have pristine === true after initialize and before the first change.

Here is what I want to get:
1. Initialize form, show no messages, pristine === true
2. Make a change, show "you have unsaved changes" message, pristine === false 
3. Submit form, show "submit succeeded" message, pristine === true
4. Make a change, show "you have unsaved changes" message, pristine === false

**Comment by iamdanthedev** [comment]
@dmitrykrylov  yes this is exactly what I tried to achieve myself. Now it's difficult to remember, but I can see in the code that I ended up having this:

```
  enableReinitialize: true,
  keepDirtyOnReinitialize: true,
```

but I can't say if it helped or not :-)

**Comment by dmitrykrylov** [comment]
@iamdanthedev yeah, just setting `keepDirtyOnReinitialize: true` has worked for me

**Comment by F-Adam-B** [comment]
I'm struggling to get a pristine state after a form submit. Here is my reduxForm() 

`const ConfigurationFormHOC = reduxForm({
  form: 'configurationForm',
  validate,
  warn,
  onSubmit: submit,
  enableReinitialize: true,
  keepDirtyOnReinitialize: true,
  onSubmitSuccess: (result, dispatch, props) => {
    dispatch(initialize('configurationForm', result, false)); // the server returns config params we don't necessarily have. Reinitialize with those values.
    dispatch(push(`/configure/${result.id}`)); // update route
  },
  onSubmitFail: (errors, dispatch, submitError, props) => {
    console.info('Submit errors:', errors);
    dispatch(configurationStatus(false));
  },
})(ConfigurationForm);`

I'm trying to mount my component after a successful submit using componentDidUpdate() with this, 

`  componentDidUpdate(prevProps, prevState, snapShot) {
    if (prevProps.submitSucceeded) {
      this.props.dispatch(initialize('configurationForm', prevProps.successfulFormValues));
      this.showSuccess = true;
    }
    if (prevProps.feasible && prevProps.pristine) {
      this.props.dispatch(configurationStatus(true));
    } else if (prevProps.feasible === false) {
      this.props.dispatch(configurationStatus(false));
    }
  }
`

The form never returns to a pristine state. Any ideas?

**Comment by IUnknown68** [comment]
Seems like the error appears only when the `initialValues` contain an attribute, that is not under control of the form, yet computed from form controlled attributes.

Example:
```js
const user = {
  firstName: 'John',
  lastName: 'Doe',
},

user.fullName = getFullname(user); // fullName is 'John Doe'
```
Change `firstName` and save. Start editing again (focusing a field is enough). `initialValues` now is:

```js
{
  firstName: 'Harry',
  lastName: 'Doe',
  fullName: 'Harry Doe',
},
```

but [`stateInitial`](https://github.com/erikras/redux-form/blob/d7d4e0b89696436fb6550ae07b94fc1b7b467f0f/src/createReduxForm.js#L1001) still is:

```js
{
  firstName: 'Harry',
  lastName: 'Doe',
  fullName: 'John Doe',
},
```

causing [`deepEqual(initialValues, stateInitial)`](https://github.com/erikras/redux-form/blob/d7d4e0b89696436fb6550ae07b94fc1b7b467f0f/src/createReduxForm.js#L1007) to return `false`.

So to fix in your application: Remove the (computed) attribute (in my case the computed `fullName`) from your `initialValues`.
